### ISSUE TITLE
[BUG] Groupby aggregation sums are more non-deterministic than expected with float32 values

### ISSUE THREAD

**Issue by beckernick** [rapidsai/cudf#3464]
Groupby sum aggregations are more non-deterministic than expected when there are float32 values in the column to be summed. Casting to float64 resolves this issue. It's hard to reproduce with only a few rows, so attaching a CSV file of 200 rows.

```python
import cudf

dtype_dict = {'0': 'int32',
 '1': 'int32',
 '2': 'int32',
 '3': 'float32'}

data = cudf.read_csv('small_clean_cols.txt', dtype=dtype_dict)

old = None
for i in range(100):
    temp = data.groupby(by=['0','1','2']).agg({'3':'sum'})
    if not old:
        old = float(temp['3'].iloc[0])
        continue
        
    current = float(temp['3'].iloc[0])
    if abs(current - old) > .0001:
        print(old, current)
        print(i)
        
    old = current
129200.9453125 129200.9609375
1
129200.9609375 129200.953125
2
129200.953125 129200.9453125
...
```

No differences here:
```python
import cudf

dtype_dict = {'0': 'int32',
 '1': 'int32',
 '2': 'int32',
 '3': 'float64'}

data = cudf.read_csv('small_clean_cols.txt', dtype=dtype_dict)

old = None
for i in range(100):
    temp = data.groupby(by=['0','1','2']).agg({'3':'sum'})
    if not old:
        old = float(temp['3'].iloc[0])
        continue
        
    current = float(temp['3'].iloc[0])
    if abs(current - old) > .0001:
        print(old, current)
        print(i)
        
    old = current
```

Data: [small_clean_cols.txt](https://github.com/rapidsai/cudf/files/3893893/small_clean_cols.txt)

**Comment by jrhemstad** [comment]
This has come up before: https://github.com/rapidsai/cudf/issues/2494

**Comment by beckernick** [comment]
Per additional discussion, the max/min of each group are:

```
max	min
0	7846.220215	0.000000
1	4794.240234	0.000000
2	14173.639650	7.500000
3	4029.479980	19.200001
4	6240.429199	0.000000
5	7062.750000	27.200001
6	3568.620117	74.229996
```

**Comment by jrhemstad** [comment]
The magnitude of these differences (~0.01) is larger than what I would expect for normal floating point round off error. Something fishy here.

**Comment by harrism** [comment]
@beckernick I want to eliminate floating point non-associativity as the cause here. So here are some experiments to try. First, can you change the printing so that you print the absolute difference no matter what the threshold is, for both fp32 and fp64 case?

1. Repeat this test with Pandas, and randomize the order of the inputs between runs? (So that the values used are always the same, just the sums are computed in random order)
2. Same as 1 but with cuDF: Original cuDF test, but randomize the order of the inputs between runs.

The reason for this is that there is no difference in the C++ code for fp32 and fp64 -- only the types used for computation.

**Comment by rgsl888prabhu** [comment]
I tried something similar with pandas with a little modification in the code
```
import pandas as pd

dtype_dict = {'0': 'int32',
 '1': 'int32',
 '2': 'int32',
 '3': 'float32'}

data = pd.read_csv('small_clean_cols.txt', dtype=dtype_dict)

old = None;
for i in range(0, 10) :
    col = data['3'].sample(n=data['3'].count(), random_state=i)
    current = col.sum()
    if (old != None):
        print (old - current)

    old = current;
```
Result:
```
0.03125
0.0
0.0
-0.03125
0.03125
0.0
0.0
-0.03125
0.0
```
After converting to `float64`
```
-5.820766091346741e-11
5.820766091346741e-11
0.0
0.0
0.0
0.0
0.0
0.0
-5.820766091346741e-11
(cudf_dev) rgsl888@onepiece:~
```

```

**Comment by jrhemstad** [comment]
@beckernick so it looks like this is just normal floating-point behavior (as expected). Can you confirm @rgsl888prabhu 's experiment looks sane to you?

**Comment by harrism** [comment]
:) I love the scientific method.  So, a couple of points.

1. floating point arithmetic is non-associative: order of operations affects the answer. Parallelizing summation results in a different order of operations compared to sequential summation. Not only that, but parallel implementations run parts of the summation concurrently, which means the order of operations is non-deterministic.  
2. This does not mean the parallel version is less accurate. In fact, parallel implementations tend to use tree-based reduction, also known as pair-wise summation. This is in fact a strategy often used in sequential summations to increase accuracy (reasons why it is more accurate left as exercise for another storytime). In fact, I ran across this SO answer: https://stackoverflow.com/a/46979587/749748, which explains that numpy (and therefore Pandas) even use pairwise summation in some circumstances.

I'm not surprised that the size of the differences between runs led you to think that the behavior is "more non-deterministic than expected", Nick. At first glance I thought the same. But I think the experiments show that totally random order of summation results in the same order of magnitude differences (even greater?).

It's impressive just how much smaller the differences are for double precision. But some hand-wavy math can explain it...  In (highly recommended) ["What Every Computer Scientist Should Know about Floating Point"](https://www.itu.dk/~sestoft/bachelor/IEEE754_article.pdf), Goldberg calculates that (for vanilla sequential summation), "Doubling the precision has the effect of squaring epsilon". Moreover, the error is relative to the values themselves, which means the total error is roughly proportional to the total sum. (See section 4.3). The following is definitely an approximation, but...

There are 200 values in the example CSV file, and their sum is ~389,000.

Single precision machine epsilon = 5.96e-08
Double precision machine epsilon = 1.11e-16

Therefore, for a sequential summation, we should expect:

Single-precision error ~= 5.96e-08 * 389000 = 2.32e-02
Double-precision error ~= 1.11e-16 * 389000 = 4.32e-11

Those numbers are very close to what we see in practice in the experiments above.

**Comment by rgsl888prabhu** [comment]
Is it because we have only 200 values and these tend to have only these 3/4 set of end values ?

**Comment by harrism** [comment]
@rgsl888prabhu I don't know.

**Comment by datametrician** [comment]
This may have been the most fascinating GitHub issue I've seen in a while :) Thank you team.

**Comment by harrism** [comment]
@rgsl888prabhu I re-read your question and realized I don't understand the question. Can you rephrase?

**Comment by rgsl888prabhu** [comment]
> @rgsl888prabhu I re-read your question and realized I don't understand the question. Can you rephrase?

@harrism The number of values in that column are 200, and adding those values in any order might result in one of those 3/4 values, which might explain why we can see only couple of different values in my result which repeat.

And I have tried only with 10 iterations, with 200 values there are lot of possibilities of different values.

**Comment by harrism** [comment]
What do you mean by 3/4 values?

**Comment by rgsl888prabhu** [comment]
3 or 4, my bad :)

**Comment by harrism** [comment]
Ah, yes. Only binary numbers are representable, so there are a discrete set of possible sums. So if we look at the values originally reported by @beckernick 's program:

```
129200.9453125 129200.9609375
129200.9609375 129200.953125
129200.953125 129200.9453125 
```

There are 3 unique values there.
```
129200.9453125 == b01000111111111000101100001111001
129200.9531250 == b01000111111111000101100001111010
129200.9609375 == b01000111111111000101100001111011
```

The error is basically in the last bits of the representation, so there is a very small number of representable binary values. I suspect if you run it enough you will also see:

```
129200.93750 == b01000111111111000101100001111000 
129200.96875 == b01000111111111000101100001111100
```

And possibly a couple of others. What values do you get? This site is handy for visualizing: 
https://www.h-schmidt.net/FloatConverter/IEEE754.html

**Comment by rgsl888prabhu** [comment]
This is for summing all the values in that column
```
389643.90625     ----  01001000101111100100000101111101
389643.875         ----  01001000101111100100000101111100
389643.9375       ---- 01001000101111100100000101111110
389643.96875     ---- 01001000101111100100000101111111
389643.84375     ---- 01001000101111100100000101111011
```
And as you said, its last couple of bits.

**Comment by harrism** [comment]
OK, I think all mysteries are explained. Thanks! :)

**Comment by beckernick** [comment]
This is super interesting. Thank you all for the deep dives on this, and thanks for the link to that PDF @harrism . Will give it a read over the weekend, as it looks a bit dense.

Since it seems like this issue is resolved for all intents and purposes, I'll close it.
